javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/RemappingUrlStreamHandler.java

```diff
@@ -0,0 +1,100 @@
+import net.bytebuddy.jar.asm.commons.ClassRemapper;
+
+class RemappingUrlStreamHandler extends URLStreamHandler {
+
+  // We need to prefix the names to prevent the gradle shadowJar relocation rules from touching
+  // them. It's possible to do this by excluding this class from shading, but it may cause issue
+  // with transitive dependencies down the line.
+  private static final ShadingRemapper remapper =
+      new ShadingRemapper(
+          rule("#io.opentelemetry.api", "#io.opentelemetry.javaagent.shaded.io.opentelemetry.api"),
+          rule(
+              "#io.opentelemetry.context",
+              "#io.opentelemetry.javaagent.shaded.io.opentelemetry.context"),
+          rule(
+              "#io.opentelemetry.extension.aws",
+              "#io.opentelemetry.javaagent.shaded.io.opentelemetry.extension.aws"),
+          rule("#java.util.logging.Logger", "#io.opentelemetry.javaagent.bootstrap.PatchLogger"),
+          rule("#org.slf4j", "#io.opentelemetry.javaagent.slf4j"));
+
+  private final JarFile delegateJarFile;
+
+  public RemappingUrlStreamHandler(URL delegateJarFileLocation) {
+    try {
+      delegateJarFile = new JarFile(new File(delegateJarFileLocation.toURI()), false);
+    } catch (URISyntaxException | IOException e) {
+      throw new IllegalStateException("Unable to read internal jar", e);
+    }
+
+    System.out.println("Delegate jar file is " + delegateJarFile.getName());
+  }
+
+  /** {@inheritDoc} */
+  protected URLConnection openConnection(URL url) {
+    System.out.println("Open connection to " + url);
+    try {
+      String file = url.getFile();
+      if ("/".equals(file)) {
+        // "/" is used as the default url of the jar
+        // This is called by the SecureClassLoader trying to obtain permissions
+
+        // nullInputStream() is not available until Java 11
+        return new InternalJarUrlConnection(url, new ByteArrayInputStream(new byte[0]), 0);
+      }
+
+      if (file.startsWith("/")) {
+        file = file.substring(1);
+      }
+      System.out.println("Looking for file " + file);
+      JarEntry entry = delegateJarFile.getJarEntry(file);
+      if (entry == null) {
+        return null;
+      }
+
+      System.out.println("Found " + entry);
+
+      InputStream is = delegateJarFile.getInputStream(entry);
+      if (file.endsWith(".class")) {
+        System.out.println("Remapping... ");
+        byte[] remappedClass = remapClassBytes(is);
+        return new InternalJarUrlConnection(
+            url, new ByteArrayInputStream(remappedClass), remappedClass.length);
+      } else {
+        System.out.println("Return as is");
+        return new InternalJarUrlConnection(url, is, entry.getSize());
+      }
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    return new InternalJarUrlConnection(url, new ByteArrayInputStream(new byte[0]), 0);
+  }
+
+  private static byte[] remapClassBytes(InputStream in) throws IOException {
+    ClassWriter cw = new ClassWriter(0);
```

Comment:
usually it is better to pass `ClassReader` to `ClassWriter` this way asm can copy methods that are not transformed directly from read to writer. Though I guess when remapping all methods get passed through a `MethodVisitor` and this optimization wouldn't work anyway.